### ISSUE TITLE
respond_to matcher fails when OUT and acts as a proxy for another object via method_missing

### ISSUE THREAD

**Issue by kgrz** [rspec/rspec-expectations#308]
**TL;DR**  `respond_to()` matcher fails when object under test has a `method_missing` directive using which, it acts as a proxy for another object that has the method defined.

Note that this issue is not the same as the one here: https://github.com/rspec/rspec-expectations/pull/183 but it's worth going through the discussion. Especially, [this article by Yehuda](http://yehudakatz.com/2010/01/02/the-craziest-fing-bug-ive-ever-seen/), originally linked there.

This is how Ruby (2.0) works:

``` ruby
class A
  def a
    "Haha. This is A"
  end
end

class B
  def b
    "Haha. This is B"
  end

  def method_missing(method)
    A.new.send(method)
  end
end

__END__

B.new.a  # => "Haha. This is A"

b = B.new
b.respond_to?(:a)  # => false. I was under the impression it returned true!
```

The [`Rspec::Matchers::BuiltIn::RespondTo`](https://github.com/rspec/rspec-expectations/blob/master/lib/rspec/matchers/built_in/respond_to.rb#L46) class uses the `respond_to?` method to figure out whether the object under test responds to the sent method or not. Due to this, the following code fails:

``` ruby
it "should actually pass as far as I'd expect" do
  expect(b).to respond_to?(:a)
end
# This fails.
```

May be, the fact that `respond_to?` returns `true` only if the method is explicitly defined on the object or on it's parents, is a legitimate design choice (since it survived two versions of Ruby). But should the same behavior be preserved in Rspec? Or should `respond_to` matcher match and return `true` in the above example?
